services/frontend/domain/usecases/process_manifest_files.usecase.go

```diff
@@ -0,0 +1,154 @@
+// corrected or deleted.
+func (u *Usecases) ProcessManifestFiles(manifestDir string) error {
+
+	manifestFiles, err := os.ReadDir(manifestDir)
+	if err != nil {
+		return fmt.Errorf("Failed to read manifest files from dir %q: %w", manifestDir, err)
+	}
+
+	configs, err := u.ContextBox.GetAllConfigs()
+	if err != nil {
+		return fmt.Errorf("Failed to retrieve manifests details from context-box: %w", err)
+	}
+
+	log.Debug().Msgf("%d configs present in database | %d configs in %v", len(configs), len(manifestFiles), manifestDir)
+
+	manifestProcessingResultsChan := make(chan *ManifestProcessingResult, len(manifestFiles))
+
+	waitGroup := sync.WaitGroup{}
+
+	for _, manifestFile := range manifestFiles {
+		waitGroup.Add(1)
+
+		// Process each of the files concurrently in a separate go-routine skipping over files for which
+		// an error occurs.
+		// By processing, we mean unmarshalling the claudie manifest
+		go func(manifestFile fs.DirEntry) {
+			defer waitGroup.Done()
+
+			manifestFilepath := filepath.Join(manifestDir, manifestFile.Name())
+
+			var (
+				rawManifestData []byte
+				manifest        manifest.Manifest
+				processingError error
+			)
+
+			defer func() {
+				manifestProcessingResultsChan <- &ManifestProcessingResult{
+					manifestName:     manifest.Name,
+					rawManifestData:  rawManifestData,
+					manifestFilepath: manifestFilepath,
+					processingError:  processingError,
+				}
+			}()
+
+			if rawManifestData, processingError = os.ReadFile(manifestFilepath); err != nil {
+				return
+			}
+
+			if processingError = yaml.Unmarshal(rawManifestData, &manifest); processingError != nil {
+				return
+			}
+
+			manifest.Validate()
+		}(manifestFile)
+	}
+
+	go func() {
+		waitGroup.Wait()
+
+		close(manifestProcessingResultsChan)
+	}()
+
+	// Collect processing results of manifest files which were processed successfully
+	for manifestProcessingResult := range manifestProcessingResultsChan {
+		// TODO: review
+		if manifestProcessingResult.processingError != nil {
+			log.Error().Msgf("Skipping over file %v due to processing error : %v", manifestProcessingResult.manifestFilepath, manifestProcessingResult.processingError)
+			continue
+		}
+
+		configs = removeConfig(configs, manifestProcessingResult.manifestName)
+
+		err = u.ContextBox.SaveConfig(
+			&pb.Config{
+				Name:     manifestProcessingResult.manifestName,
+				Manifest: string(manifestProcessingResult.rawManifestData),
+			},
+		)
+		if err != nil {
+			log.Error().Msgf("Skipped saving config %v due to error : %v", manifestProcessingResult.manifestName, err)
+			continue
+		}
+
+		log.Info().Msgf("Details of manifest file %s has been saved to context-box database", manifestProcessingResult.manifestFilepath)
+	}
+
+	// threadSafeMap is a go-routine safe map that stores id of configs that are being currently deleted
+	// to avoid having multiple go-routines deleting the same configs from MongoDB (of contextBox microservice).
+	var threadSafeMap sync.Map
```

Comment:
Do we really need this map? If we iterate over `configs`, how can get a single config get picked up twice?

Reply:
In the original codebase, the **server** struct contains a property called **deletingConfigs** which is a `go-routine safe map` -
```go
type server struct {
    ...
    // deletingConfigs is a go-routine safe map that
    // stores configs that are being currently deleted
    // to avoid having multiple go-routines deleting the
    // same config from the database.
    deletingConfigs sync.Map
    ...
}
```

Those manifest files which were deleted from the directory, needs to be deleted from the context-box DB as well. Here is the implementation in  the original codebase -
```go
for _, config := range configs.Configs {
    if _, ok := s.deletingConfigs.Load(config.Id); ok {
        continue
    }
    s.deletingConfigs.Store(config.Id, nil)
    go func(config *pb.Config) {
        ...
	s.deletingConfigs.Delete(config.Id)
    }(config)
}
```

You can see the **go-routine safe map** is used here!

Reply:
In the current version there is a `sync.Map` because if you would create 2 or 3 secrets one behind the other multiple go-routines would be spawned that would process the stored manifests.

https://github.com/berops/claudie/blob/master/services/frontend/server.go#L152-L159

The `sync.Map` is just a guard to be sure that the deletion process is started only once and not multiple times.

Whereas in this refactored version https://github.com/berops/claudie/blob/f8a04437c0b15fd582b4aa9d097ee848cfb024bd/services/frontend/domain/usecases/process_manifest_files.usecase.go#L129-L131

There is no guard againts that since the map is defined in the same function where the configs are being processed